src/main/java/com/conveyal/gtfs/loader/PatternReconciliation.java

```diff
@@ -0,0 +1,455 @@
+        reconcilePattern(tablePrefix, patternId, patterns, connection, PatternType.STOP);
+    }
+
+    /**
+     * Reconcile pattern locations.
+     */
+    public static void reconcilePatternLocations(
+        String tablePrefix,
+        String patternId,
+        List<PatternLocation> patternLocations,
+        Connection connection
+    ) throws SQLException {
+        List<Pattern> patterns = patternLocations.stream().map(Pattern::new).collect(Collectors.toList());
+        reconcilePattern(tablePrefix, patternId, patterns, connection, PatternType.LOCATION);
+    }
+
+    /**
+     * We assume only one stop time has changed, either it's been removed, added or moved. The only other case that is
+     * permitted is adding a set of stops/locations to the end of the original list. These conditions are evaluated by
+     * simply checking the lengths of the original and new pattern stops/locations (and ensuring that stop IDs remain
+     * the same where required). If the change to pattern stops/locations does not satisfy one of these cases, fail the
+     * update operation.
+     *
+     */
+    private static void reconcilePattern(
+        String tablePrefix,
+        String patternId,
+        List<Pattern> patterns,
+        Connection connection,
+        PatternType patternType
+    ) throws SQLException {
+
+        LOG.info("Reconciling pattern for pattern ID={} for pattern type={}", patternId, patternType);
+        if (patterns.isEmpty()) return;
+
+        List<String> tripsForPattern = getTripIdsForPatternId(tablePrefix, patternId, connection);
+        if (tripsForPattern.size() == 0) {
+            // If there are no trips for the pattern, there is no need to reconcile stop times to modified patterns.
+            // This permits the creation of patterns without stops, reversing the stops on existing patterns, and
+            // duplicating patterns.
+            // For new patterns, this short circuit is required to prevent the transposition conditional check from
+            // throwing an IndexOutOfBoundsException when it attempts to access index 0 of a list with no items.
+            return;
+        }
+
+        List<String> originalPatternIds = getOriginalPatternIds(tablePrefix, patternId, patternType, connection);
+
+        // Prepare SQL fragment to filter for all stop times for all trips on a certain pattern.
+        String joinToTrips = String.format("%s.trips.trip_id = %s.stop_times.trip_id AND %s.trips.pattern_id = '%s'",
+            tablePrefix, tablePrefix, tablePrefix, patternId);
+
+        // ADDITIONS (IF DIFF == 1)
+        if (originalPatternIds.size() == patterns.size() - 1) {
+            // We have an addition; find it.
+            int differenceLocation = -1;
+            for (int i = 0; i < patterns.size(); i++) {
+                if (differenceLocation != -1) {
+                    // we've already found the addition
+                    if (i < originalPatternIds.size() &&
```

Comment:
Combining produces an incorrect result. If the combined statement is false the else if is triggered updating the `differenceLocation` parameter. If adding a stop to the beginning, instead of the `differenceLocation` being zero it is three which results in the wrong stop sequences being updated.

Reply:
Ok it's because of the `else if` that follows.